sdk/identity/azure-identity/azure/identity/_credentials/default.py

```diff
@@ -108,6 +108,9 @@ def get_token(self, *scopes, **kwargs):
         :raises ~azure.core.exceptions.ClientAuthenticationError: authentication failed. The exception has a
           `message` attribute listing each authentication attempt and its error message.
         """
+        if self._successful_credential:
+            return self._successful_credential.get_token(*scopes, **kwargs)
```

Comment:
Let's say we have 3 credentials in the chain.
First time, #1 failed, #2 succeeded.
Second time, #2 failed, do we want to try #3?
What's the expected behavior?

Reply:
Today credential N's success is irrelevant because `ChainedTokenCredential` (and by inheritance `DefaultAzureCredential`) always tries all credentials until one returns a token or raises `ClientAuthenticationError`. This is fine for `ChainedTokenCredential` but not what we want from `DefaultAzureCredential`.

The desired behavior of `DefaultAzureCredential` is not to iterate the chain again after a credential returns a token, because doing so only wastes cycles and introduces the possibility of an unexpected change of authentication method.

Reply:
Yes. But the concern is by caching the last successful credential, we change the behavior.
The old behavior: if last successful credential fails, we try next
New behavior: if last successful credential fails, we return fail.
Is this expected?

Reply:
> 
> The old behavior: if last successful credential fails, we try next
> New behavior: if last successful credential fails, we return fail.

We must be more specific here. It depends on how the credential fails:
- `CredentialUnavailableError`: try the next credential
- `ClientAuthenticationError`: raise

Credentials generally don't raise `CredentialUnavailableError` after having succeeded, which means "old" and "new" behavior are equivalent in most cases. `AzureCliCredential` could do it, if someone uninstalled the CLI or the shell between calls to `get_token`. However, that credential is at the end of the chain unless the user opted in to interactive authentication, so the behavior is still typically unchanged.

Back to your question:
> Is this expected?

Yes. Consider the alternative: how surprised would you be if `DefaultAzureCredential` authenticated with the CLI the first time, then popped a browser for you to log in the second time?

Reply:
I am fine with it. Just want to make sure this is the design.